### DISCUSSION TITLE
How does SealedSecret know which encryption cert to use for unsealing a secret?

### DISCUSSION THREAD

**Discussion by chinmaya-n** [bitnami-labs/sealed-secrets#1191] [Q&A] [answered]
How does it know? Or does it just brute force and cycle through all the encryption certs until one successfully unseals it?

**Comment by tewfik-ghariani** [comment, marked as answer]
At the moment, all possible private keys are being tried and there is a ToDo item to change that

> // TODO(mkm): use the key fingerprint encoded in ciphertext (if present) instead of trying all the possible keys

https://github.com/bitnami-labs/sealed-secrets/blob/bb942f1acfd6791ce4da535fe263dca88c664e76/pkg/crypto/crypto.go#L76

That `HybridDecrypt` function is called by the `Unseal` fct: https://github.com/bitnami-labs/sealed-secrets/blob/bb942f1acfd6791ce4da535fe263dca88c664e76/pkg/apis/sealedsecrets/v1alpha1/sealedsecret_expansion.go#L278

And the `Unseal` function is called after looping through all private keys in the key registry: https://github.com/bitnami-labs/sealed-secrets/blob/bb942f1acfd6791ce4da535fe263dca88c664e76/pkg/controller/controller.go#L499

**Reply by mkmik** [reply]
yeah ideally the encyrpted data should contain the id of the keypair used to encrypt so that it can be used to quickly find the right one to decrypt (and also to provide useful troubleshooting answers in case the key is not found!).

There should be an open task to do it. All we need is somebody with some time to implement it. Feel free to tag me for a review
